Improve hash mixing in FST's double-barrel LRU hash

shubhamvishu opened the pull request:
### Description

 Addresses #12704. Below is the comment that inspired this ([link](https://github.com/apache/lucene/pull/12633#discussion_r1366847986)),


```
Instead, we should multiply with the gold constant BitMixer#PHI_C64 (make it public).
This really makes a difference in the evenness of the value distribution. This is one of the secrets of the HPPC hashing. By applying this, we get multiple advantages:

  * lookup should be improved (less hash collision)
  * we can try to rehash at 3/4 occupancy because the performance should not be impacted until this point.
  * in case of hash collision, we can lookup linearly with a pos = pos + 1 instead of quadratic probe (lines 95 and 327); this may avoid some mem cache miss.
  
(same for the other hash method)
```

Closes #12704 
<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

shubhamvishu committed:
Improve hash mixing in FST's double-barrel LRU hash

shubhamvishu committed:
Add comments and changes entry

mikemccand commented on lucene/CHANGES.txt:
```diff
@@ -190,6 +190,8 @@ Improvements
 * GITHUB#12705, GITHUB#12705: Improve handling of NullPointerException and IllegalStateException
   in MMapDirectory's IndexInputs.  (Uwe Schindler, Michael Sokolov)
 
+* GITHUB#12716: Improve hash mixing in FST's double-barrel LRU hash. (Shubham Chaudhary)
```
Maybe also add something more end-user friendly like `, making FST compilation faster and more RAM-efficient` or so?

shubhamvishu replied:
Changed in the new revision!

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/fst/NodeHash.java:
```diff
@@ -182,8 +183,9 @@ private long hash(FSTCompiler.UnCompiledNode<T> node) {
-    return h;
+    // Here we multiply the hash with the gold constant BitMixer#PHI_C64
+    // This makes a real difference in the evenness of the value distribution
+    return h * BitMixer.PHI_C64;
```
Could you also add a link that @dweiss provided in [this comment](https://github.com/apache/lucene/issues/12704#issuecomment-1774156970)?  I always freak out a bit when I see magical constants and I don't understand their magic.  [That link](https://softwareengineering.stackexchange.com/questions/402542/where-do-magic-hashing-constants-like-0x9e3779b9-and-0x9e3779b1-come-from) goes into a nice little complex history lesson...).  Or perhaps put the link on the `BitMixer.PHI_C64` as a comment?

shubhamvishu replied:
I have added the link on the constant. Its was better to add the link at one place(where we declare) than on all usages.

mikemccand reviewed (commented):
Thanks @shubhamvishu!  This looks great -- I left some small comments.

Could you also change the probe from quadratic (what it is now) to a simple linear probe (just test each `+1` (wrapping at the hash table size) slot on insert collision)?

Second, could you change the rehash ratio from 2/3 -> 3/4?  The better mixing should mean we can run a tighter ship here.

I'm happy to help test the impact to FST construction time -- I'll just run the new `IndexToFST` tool in `luceneutil` on a nightly `wikimediumall` index.  That's a decent stress test of FST building (creates an FST ~350 MB or more, depending on how much RAM you are willing to spend while compiling it).

With [this upcoming issue](https://github.com/apache/lucene/issues/12714) (which will add RAM inefficiency!) it's even more important to make this hash as RAM efficient as possible.

shubhamvishu committed:
Address comments

shubhamvishu commented:
Thanks for the quick review @mikemccand! I have addressed the comments in the new revision.

> Could you also change the probe from quadratic (what it is now) to a simple linear probe (just test each +1 (wrapping at the hash table size) slot on insert collision)?

> Second, could you change the rehash ratio from 2/3 -> 3/4? The better mixing should mean we can run a tighter ship here.

Addressed both in the new revision

> I'm happy to help test the impact to FST construction time -- I'll just run the new IndexToFST tool in luceneutil on a nightly wikimediumall index. That's a decent stress test of FST building (creates an FST ~350 MB or more, depending on how much RAM you are willing to spend while compiling it).

Thank you so much for the help Mike ! I have never run the fst benchmark but seems like its straightforward java script?. I could give it a try as well (sometime this week) if you don't beat me to it.

shubhamvishu committed:
Merge branch 'main' into GH-12704

mikemccand commented:
> Thank you so much for the help Mike !

Thank you!

> I have never run the fst benchmark but seems like its straightforward java script?. I could give it a try as well (sometime this week) if you don't beat me to it.

It is fairly simple.  It's build "script" lives as a simple java comment inside it :)  The fun part is building a full `wikimediumall` index for it to pull the terms from.

mikemccand commented:
OK I ran it twice on `main`:
```
    saved FST to "fst.bin": 294815624 bytes; 59.874 sec
    saved FST to "fst.bin": 294815624 bytes; 60.255 sec
```

And twice with this PR:

```
    saved FST to "fst.bin": 294815624 bytes; 69.026 sec
    saved FST to "fst.bin": 294815624 bytes; 70.082 sec
```

Net net ~ 16.7% slower (~60s -> ~70s), maybe more than I expected.  Should we try just the better mixing but keep the rehash trigger at 2/3, maybe?

bruno-roustant commented:
Oh, the numbers are disappointing. I expected to be both a little more compact and little faster.
I wonder what is the cause, the rehash threshold, the linear scan, or the multiplication itself?

The benchmark is for one fst size. Often when benchmarking hash map, I saw the curve make stages when the size increased. But I doubt a stage would explain such a difference.

shubhamvishu commented:
@mikemccand @bruno-roustant  So I also ran the FST construction time benchmarks on `wikimediumall` index using `IndexToFST` over a couple of combinations of tweakable parameters to see the effect. Following are the results :

| Golden ratio Bit mixing | Rehash ratio (`2/3`) | Rehash ratio (`3/4`) | Quadratic Probing | Linear Probing | No. of iterations | Avg Time (seconds) | Comments |
|:-----------------------:|:------------------:|:------------------:|:-----------------:|:--------------:|-------------------|:------------------:|:--------:|
|            No           |         Yes        |          -         |        Yes        |        -       |         10        |       89.5471      | **Baseline** |
|           Yes           |         Yes        |          -         |        Yes        |        -       |         10        |       89.6913      |          |
|           Yes           |         Yes        |          -         |         -         |       Yes      |         10        |       89.8202      |          |
|           Yes           |          -         |         Yes        |        Yes        |        -       |         10        |        98.15       |          |
|           Yes           |          -         |         Yes        |         -         |       Yes      |         10        |      103.4483      |    **PR**    |

&nbsp;
Then I also tried rehash ratio of `1/2` (just to explore ahead) to see its impact, that seem to have slightly better numbers (best amongst the tried ones).

| Golden ratio Bit mixing | Rehash ratio (`1/2`) | Quadratic Probing | Linear Probing | No. of iterations | Avg Time (seconds) | Comments |
|:-----------------------:|:------------------:|:-----------------:|:--------------:|-------------------|:------------------:|:--------:|
|           Yes           |         Yes        |        Yes        |        -       |         10        |       89.1806      |          |
|           Yes           |         Yes        |         -         |       Yes      |         10        |       89.2009      |          |

I'm not aware of the rationale behind choosing rehash ratio of `2/3`, but should we try making it `1/2`? That might have few consequences which I am unaware of. Keen to know your thoughts on this. Thanks!

bruno-roustant commented:
Thank you @shubhamvishu for these experiments. The table answers exactly to the questions.
And it means actually there is no point to change the bit mixing, since it does not bring any improvement.

Making the rehash threshold at 1/2 would means the hash structure would occupy more memory on average since it would allocate an "array" and would use it between 25% to 50% occupancy. With the 2/3 threshold, it uses it between 33% and 66% occupancy. So staying at 2/3 is better.

Sometimes an appealing idea is actually a false good idea. And it's great to have these experiments to avoid the pitfall!

shubhamvishu commented:
@bruno-roustant Absolutely! I totally agree with your point.

mikemccand commented:
Wow, these are interesting results.  Thanks @shubhamvishu.  It's curious that, at 3/4 rehash ratio, linear probing is quite a bit slower (103.4 s) than quadratic (98.1 s).

The fact that the FST compilation time did not change much between 2/3 rehash trigger and 1/2 rehash trigger makes me wonder if there is a better trigger somewhere even beyond 2/3 but before 3/4.  But I don't think this is urgent to explore ... let's keep the 2/3 ratio for now?

Thanks for all the testing -- indeed not every idea pans out, and it's great that that happens.  If every idea panned out then we are not  trying hard enough to come up with innovative ideas :)

mikemccand commented:
@shubhamvishu can we close this one?  Any other things to try?  For some reason, FST building of enwiki terms seems not to like this magical hashing...

shubhamvishu commented:
> @shubhamvishu can we close this one? Any other things to try? 

Sure @mikemccand ! Maybe we could just try a rehash value between 2/3 and 3/4 as you mentioned earlier(how about 17/24? the mid value of these two) and close this post that since we are at it now. I'll run it and share the results for that soon.

&nbsp;
> For some reason, FST building of enwiki terms seems not to like this magical hashing...

Indeed!💯

shubhamvishu commented:
Ok so I ran the test with rehash value of 17/24 which is between 2/3 and 3/4. Here are the results:

| Golden ratio Bit  mixing  | Rehash ratio  (2/3) | Rehash ratio  (17/24) | Quadratic Probing | No. of iterations | Avg Time  (seconds) |   Comments   |
|:-------------------------:|:-------------------:|:---------------------:|:-----------------:|:-----------------:|:-------------------:|:------------:|
|             No            |         Yes         |           -           |        Yes        |         10        |       103.4224      |   Baseline   |
|             No            |          -          |          Yes          |        Yes        |         10        |       103.0535      |   Candidate  |

Results are almost the same(maybe just noise?). @mikemccand please feel free to close this PR if this looks not worth pursuing to you. Thanks!

dweiss commented:
The fact hash key remixing doesn't improve the situation is not necessarily a sign that it's somehow wrong - it means hash keys are distributed evenly already (which is good). Remixing adds this extra layer of confidence that even if bad hash functions are present on input, it'll still redistribute those "bad" original keys fairly uniformly.

I'm surprised linear probing doesn't yield an improvement. Perhaps it's not a significant factor because of other load? Hard to say. Anyway, no need to make things more complicated if they don't yield a visible improvement.

shubhamvishu commented:
> I'm surprised linear probing doesn't yield an improvement. Perhaps it's not a significant factor because of other load? Hard to say. Anyway, no need to make things more complicated if they don't yield a visible improvement.

I totally resonate with you @dweiss on this! I'll close this PR now. Thanks everyone for looking and sharing your thoughts.😀

mikemccand commented:
Thanks @shubhamvishu and @dweiss and @bruno-roustant.  Hashing is fun and hard :)
